### PR TITLE
LPD-23786 Adapt Object Validation to support a new custom object validation engine

### PR DESCRIPTION
**Pull request by pedro-oliveira446**
[LPD-23786](https://liferay.atlassian.net/browse/LPD-23786)

In this PR, I've introduced a new type of key called 'javaDelegate#'. All custom validations that utilize this name as a prefix in the key will now receive both the originalEntryDTO and the entryDTO. To validate this enhancement, I've implemented a new engine for Bookmarks object definition. Here is the commit https://github.com/pedro-oliveira446/liferay-portal/pull/76/commits/2daf7c2953bb3103704f7fc26893c8af4b969630

### CODE REVIEW

File: modules/apps/object/object-api/src/main/java/com/liferay/object/constants/ObjectValidationRuleConstants.java [hunk 1]
```diff
@@ -16,6 +16,9 @@ public class ObjectValidationRuleConstants {
 
 	public static final String ENGINE_TYPE_GROOVY = "groovy";
 
+	public static final String ENGINE_TYPE_JAVA_DELEGATE_PREFIX =
```
Comment (p-albuquerque) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/constants/ObjectValidationRuleConstants.java:19]: Why the engine name can't be just javaDelegate?
Why it's needed append the asset name to compose the key?

**Review by p-albuquerque — COMMENTED** [review]
This approach of group the engines which have this specific type by this prefix was alligned and designed together with someone else?

**Comment by pedro-oliveira446** [comment]
Yes,
I aligned this with Gabriel, Camacho, and Feliphe.

**Comment by pedro-oliveira446** [comment]
ci:test:sf

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java [hunk 2]
```diff
@@ -632,7 +646,11 @@ private void _validate(
 			!(objectValidationRuleEngine instanceof
 				FunctionObjectValidationRuleEngineImpl ||
 			  objectValidationRuleEngine instanceof
-				  UniqueCompositeKeyObjectValidationRuleEngineImpl)) {
+				  UniqueCompositeKeyObjectValidationRuleEngineImpl ||
+			  StringUtil.startsWith(
+				  engine,
+				  ObjectValidationRuleConstants.
+					  ENGINE_TYPE_JAVA_DELEGATE_PREFIX))) {
```
[the remark this answers is not in the document]
Comment (pedro-oliveira446) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java:653]: Hey,
there is already a negation before the first parenthesis.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java [hunk 3]
```diff
@@ -3221,6 +3231,85 @@ public void testUpdateObjectEntry() throws Exception {
...
+					null, objectValidationRuleResults.get(0));
+			}
+		}
+		finally {
+			_objectValidationRuleLocalService.deleteObjectValidationRule(
+				objectValidationRule);
+
+			serviceRegistration.unregister();
+		}
+	}
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java:3311]: Hey @pedro-oliveira446 can we do this?

```suggestion
	public void testUpdateObjectEntryWithJavaDelegateObjectValidationRule()
		throws Exception {
			
		Bundle bundle = FrameworkUtil.getBundle(ObjectEntryLocalServiceTest.class);

		BundleContext bundleContext = bundle.getBundleContext();

		String key = ObjectValidationRuleConstants.ENGINE_TYPE_JAVA_DELEGATE_PREFIX +
				RandomTestUtil.randomString();

		ServiceRegistration<ObjectValidationRuleEngine> serviceRegistration = 
		bundleContext.registerService(
				ObjectValidationRuleEngine.class,
				new TestObjectValidationRuleEngine(
					TestPropsValues.getCompanyId(),
					Collections.singletonList(_objectDefinition.getName()),
					key), null);

		try {

			ObjectValidationRule objectValidationRule = _addObjectValidationRule(
				key,
				LocalizedMapUtil.getLocalizedMap(RandomTestUtil.randomString()),
				"");

			ObjectEntry objectEntry = _addObjectEntry(
				HashMapBuilder.<String, Serializable>put(
					"emailAddressRequired", "bob@liferay.com"
				).put(
					"listTypeEntryKeyRequired", "listTypeEntryKey1"
				).build());

			_objectEntryLocalService.updateObjectEntry(
					TestPropsValues.getUserId(), objectEntry.getObjectEntryId(),
					HashMapBuilder.<String, Serializable>put(
						"emailAddressRequired", "john@liferay.com"
					).put(
						"listTypeEntryKeyRequired", "listTypeEntryKey1"
					).build(),
					ServiceContextTestUtil.getServiceContext());
		}
	        finally {
			serviceRegistration.unregister();
		}
	}
```

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java [hunk 4]
```diff
@@ -3979,4 +4068,90 @@ private void _testUpdateObjectEntryObjectStateTransitions()
...
+					originalEntryValues.get("emailAddressRequired"));
+			}
+
+			if (StringUtil.equals(emailAddressRequired, "john@liferay.com") &&
+				StringUtil.equals(oldEmailAddressRequired, "bob@liferay.com")) {
+
+				return HashMapBuilder.<String, Object>put(
+					"validationCriteriaMet", false
+				).build();
+			}
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java:4115]: Hey @pedro-oliveira446 here let's replace this with two assertions:

````java
Assert.assertEquals(emailAddressRequired, "john@liferay.com");
Assert.assertEquals(oldEmailAddressRequired, "bob@liferay.com");
````

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

**Review by MarinhoFeliphe — APPROVED** [review]
LGTM!! :tada:

**Comment by MarinhoFeliphe** [comment]
ci:test:relevant

**Comment by MarinhoFeliphe** [comment]
ci:test:object

**Comment by pedro-oliveira446** [comment]
ci:forward:force

**Pull request closed without merging**